src/main/java/org/jetbrains/bsp/bazel/server/DiagnosticsDispatcher.java

```diff
@@ -0,0 +1,134 @@
+    this.bspClient = bspClient;
+  }
+
+  public void collectDiagnostics(
+      Map<Uri, List<PublishDiagnosticsParams>> filesToDiagnostics,
+      BuildTargetIdentifier target,
+      String diagnosticsLocation)
+      throws IOException {
+    Diagnostics.TargetDiagnostics targetDiagnostics =
+        Diagnostics.TargetDiagnostics.parseFrom(Files.readAllBytes(Paths.get(diagnosticsLocation)));
+
+    for (Diagnostics.FileDiagnostics fileDiagnostics : targetDiagnostics.getDiagnosticsList()) {
+      LOGGER.info("Inserting diagnostics for path: {}", fileDiagnostics.getPath());
+
+      filesToDiagnostics.put(
+          getUriForPath(fileDiagnostics.getPath()), convertDiagnostics(target, fileDiagnostics));
+    }
+  }
+
+  public void emitDiagnostics(
+      Map<Uri, List<PublishDiagnosticsParams>> filesToDiagnostics, BuildTargetIdentifier target) {
+    for (SourceItem source : bspServer.getCachedBuildTargetSources(target)) {
+      Uri sourceUri = Uri.fromFileUri(source.getUri());
+      if (!filesToDiagnostics.containsKey(sourceUri)) {
+        addSource(filesToDiagnostics, sourceUri, target);
+      }
+
+      if (bspClient != null) {
+        for (List<PublishDiagnosticsParams> values : filesToDiagnostics.values()) {
+          for (PublishDiagnosticsParams param : values) {
+            bspClient.onBuildPublishDiagnostics(param);
+          }
+        }
+      }
+    }
+  }
+
+  private void addSource(
+      Map<Uri, List<PublishDiagnosticsParams>> filesToDiagnostics,
+      Uri sourceUri,
+      BuildTargetIdentifier target) {
+    PublishDiagnosticsParams publishDiagnosticsParams =
+        new PublishDiagnosticsParams(
+            new TextDocumentIdentifier(sourceUri.toString()), target, new ArrayList<>(), true);
+
+    filesToDiagnostics.put(sourceUri, Lists.newArrayList(publishDiagnosticsParams));
+  }
+
+  private List<PublishDiagnosticsParams> convertDiagnostics(
+      BuildTargetIdentifier target, Diagnostics.FileDiagnostics request) {
+    List<Diagnostic> diagnostics =
+        request.getDiagnosticsList().stream()
+            .map(this::convertDiagnostic)
+            .collect(Collectors.toList());
+
+    PublishDiagnosticsParams publishDiagnosticsParams =
+        new PublishDiagnosticsParams(
+            new TextDocumentIdentifier(getUriForPath(request.getPath()).toString()),
+            target,
+            diagnostics,
+            true);
+
+    return Lists.newArrayList(publishDiagnosticsParams);
+  }
+
+  private Diagnostic convertDiagnostic(Diagnostics.Diagnostic diagProto) {
+    Optional<DiagnosticSeverity> severity =
+        Optional.ofNullable(CONVERTED_SEVERITY.get(diagProto.getSeverity()));
+
+    Position startPosition =
+        new Position(
+            diagProto.getRange().getStart().getLine(),
+            diagProto.getRange().getStart().getCharacter());
+    Position endPosition =
+        new Position(
+            diagProto.getRange().getEnd().getLine(), diagProto.getRange().getEnd().getCharacter());
```

Comment:
add new line after `,`

Reply:
This google formatting does that XD (splits _startPosition_ and takes back _endPosition_ into one line).